File: backend/photos.py
"""Stock photo lookup for recipes that don't come with one.

AI-generated recipes have no image, so we search Openverse — openly licensed
images, no API key, no signup, which keeps this inside the "nothing paid" rule.
Results are always presented as *a photo of a similar dish*, never as a photo
of the actual recipe, and the CC attribution travels with them.
"""
import requests

API = "https://api.openverse.org/v1/images/"
UA = "foodify/0.1 (personal meal planner)"
PER_PAGE = 8


def _license_label(code: str, version: str) -> str:
    """'by' + '2.0' -> 'CC BY 2.0'; 'cc0' is already prefixed, so don't double it."""
    code = (code or "").strip().upper()
    if not code:
        return ""
    prefix = "" if code.startswith("CC") else "CC "
    return f"{prefix}{code} {version or ''}".strip()


def _clean(item: dict) -> dict | None:
    url = item.get("url")
    if not url:
        return None
    return {
        "url": url,
        "thumbnail": item.get("thumbnail") or url,
        "title": (item.get("title") or "").strip()[:200],
        "creator": (item.get("creator") or "").strip()[:120],
        "license": _license_label(item.get("license") or "", item.get("license_version") or ""),
        "attribution": (item.get("attribution") or "").strip()[:400],
        "source_url": item.get("foreign_landing_url") or "",
    }


def search(query: str, page: int = 1) -> list[dict]:
    """Photos of a dish. Returns [] rather than raising — a missing photo is a
    cosmetic problem and must never fail the request that wanted one."""
    query = (query or "").strip()
    if not query:
        return []
    try:
        response = requests.get(
            API,
            params={
                "q": query,
                "page": max(1, page),
                "page_size": PER_PAGE,
                # photos of food, not diagrams or clip art
                "category": "photograph",
                "mature": "false",
            },
            headers={"User-Agent": UA},
            timeout=12,
        )
        if response.status_code != 200:  # includes the anonymous rate limit
            return []
        results = response.json().get("results") or []
    except Exception:  # noqa: BLE001 - network/parse issues degrade to "no photo"
        return []

    return [c for c in (_clean(r) for r in results) if c]

File: frontend/src/lib/photos.ts
/**
 * Stock photo lookup for recipes that don't come with one.
 *
 * AI-generated recipes have no image, so we search Openverse — openly licensed
 * images, no API key, no signup, which keeps this inside the "nothing paid"
 * rule. Openverse serves CORS headers, so the browser calls it directly and no
 * proxy is involved.
 *
 * Results are always presented as *a photo of a similar dish*, never as a photo
 * of the actual recipe, and the CC attribution travels with them.
 */
import type { StockPhoto } from "./types"

const API = "https://api.openverse.org/v1/images/"
const PER_PAGE = 8

/** 'by' + '2.0' -> 'CC BY 2.0'; 'cc0' is already prefixed, so don't double it. */
function licenseLabel(code: string, version: string): string {
    const upper = (code || "").trim().toUpperCase()
    if (!upper) return ""
    const prefix = upper.startsWith("CC") ? "" : "CC "
    return `${prefix}${upper} ${version || ""}`.trim()
}

interface OpenverseResult {
    url?: string
    thumbnail?: string
    title?: string
    creator?: string
    license?: string
    license_version?: string
    attribution?: string
    foreign_landing_url?: string
}

function clean(item: OpenverseResult): StockPhoto | null {
    if (!item.url) return null
    return {
        url: item.url,
        thumbnail: item.thumbnail || item.url,
        title: (item.title || "").trim().slice(0, 200),
        creator: (item.creator || "").trim().slice(0, 120),
        license: licenseLabel(item.license || "", item.license_version || ""),
        attribution: (item.attribution || "").trim().slice(0, 400),
        source_url: item.foreign_landing_url || "",
    }
}

/**
 * Photos of a dish. Returns [] rather than throwing — a missing photo is a
 * cosmetic problem and must never fail whatever wanted one.
 */
export async function searchPhotos(query: string, page = 1): Promise<StockPhoto[]> {
    const q = (query || "").trim()
    if (!q) return []

    const params = new URLSearchParams({
        q,
        page: String(Math.max(1, page)),
        page_size: String(PER_PAGE),
        // photos of food, not diagrams or clip art
        category: "photograph",
        mature: "false",
    })

    try {
        const response = await fetch(`${API}?${params}`)
        if (!response.ok) return [] // includes the anonymous rate limit
        const data = await response.json()
        const results: OpenverseResult[] = data.results ?? []
        return results.map(clean).filter((p): p is StockPhoto => p !== null)
    } catch {
        // offline, or the request was blocked — degrade to "no photo"
        return []
    }
}
